src/objective/cuda/cuda_rank_objective.cu

```diff
@@ -0,0 +1,382 @@
+        // regular the delta_pair_NDCG by score distance
+        if (norm && best_score != worst_score) {
+          delta_pair_NDCG /= (0.01f + fabs(delta_score));
+        }
+        // calculate lambda for this pair
+        double p_lambda = 1.0f / (1.0f + exp(sigmoid * delta_score));
+        double p_hessian = p_lambda * (1.0f - p_lambda);
+        // update
+        p_lambda *= -sigmoid * delta_pair_NDCG;
+        p_hessian *= sigmoid * sigmoid * delta_pair_NDCG;
+        atomicAdd_block(shared_lambdas + low, -static_cast<score_t>(p_lambda));
+        atomicAdd_block(shared_hessians + low, static_cast<score_t>(p_hessian));
+        atomicAdd_block(shared_lambdas + high, static_cast<score_t>(p_lambda));
+        atomicAdd_block(shared_hessians + high, static_cast<score_t>(p_hessian));
+        // lambda is negative, so use minus to accumulate
+        thread_sum_lambdas -= 2 * p_lambda;
+      }
+    }
+    atomicAdd_block(&sum_lambdas, thread_sum_lambdas);
+    __syncthreads();
+    if (norm && sum_lambdas > 0) {
+      const double norm_factor = log2(1 + sum_lambdas) / sum_lambdas;
+      if (threadIdx.x < static_cast<unsigned int>(query_item_count)) {
+        cuda_out_gradients_pointer[threadIdx.x] = static_cast<score_t>(shared_lambdas[threadIdx.x] * norm_factor);
+        cuda_out_hessians_pointer[threadIdx.x] = static_cast<score_t>(shared_hessians[threadIdx.x] * norm_factor);
+      }
+      if (MAX_ITEM_GREATER_THAN_1024) {
+        if (query_item_count > 1024) {
+          const unsigned int threadIdx_x_plus_1024 = threadIdx.x + 1024;
+          if (threadIdx_x_plus_1024 < static_cast<unsigned int>(query_item_count)) {
+            cuda_out_gradients_pointer[threadIdx_x_plus_1024] = static_cast<score_t>(shared_lambdas[threadIdx_x_plus_1024] * norm_factor);
+            cuda_out_hessians_pointer[threadIdx_x_plus_1024] = static_cast<score_t>(shared_hessians[threadIdx_x_plus_1024] * norm_factor);
+          }
+        }
+      }
+    } else {
+      if (threadIdx.x < static_cast<unsigned int>(query_item_count)) {
+        cuda_out_gradients_pointer[threadIdx.x] = static_cast<score_t>(shared_lambdas[threadIdx.x]);
+        cuda_out_hessians_pointer[threadIdx.x] = static_cast<score_t>(shared_hessians[threadIdx.x]);
+      }
+      if (MAX_ITEM_GREATER_THAN_1024) {
+        if (query_item_count > 1024) {
+          const unsigned int threadIdx_x_plus_1024 = threadIdx.x + 1024;
+          if (threadIdx_x_plus_1024 < static_cast<unsigned int>(query_item_count)) {
+            cuda_out_gradients_pointer[threadIdx_x_plus_1024] = static_cast<score_t>(shared_lambdas[threadIdx_x_plus_1024]);
+            cuda_out_hessians_pointer[threadIdx_x_plus_1024] = static_cast<score_t>(shared_hessians[threadIdx_x_plus_1024]);
+          }
+        }
+      }
+    }
+    __syncthreads();
+  }
+}
+
+template <data_size_t NUM_RANK_LABEL>
+__global__ void GetGradientsKernel_LambdarankNDCG_Sorted(
```

Comment:
differences compared with non sorted one?

Reply:
`GetGradientsKernel_LambdarankNDCG_Sorted` does not sort the items. `GetGradientsKernel_LambdarankNDCG` sorts the items in shared memory. These two are for different numbers of items in a query. When a query contains too many items, the sorting cannot be performed in shared memory. Thus a sorting is done separately, and then `GetGradientsKernel_LambdarankNDCG_Sorted ` is used.